### ISSUE TITLE
Enough stderr output in a run() call makes it hang forever when set_no_tty=1

### ISSUE THREAD

**Issue by mbroadhead** [RexOps/Rex#756]
I've found an issue that when a script outputs too much stderr, the call hangs forever under OpenSSH.
$ rex -v
(R)?ex 1.3.2

Here is my Rexfile:
```perl
use Rex -feature => '0.42';

desc <<'EOF';
Print X chars to stderr
  Options:
    --count=INTEGER
EOF
task "print_chars_to_stderr" => sub {
    my $opt = shift;
    my $count = $opt->{count} || 1;
    Rex::Logger::info( "connection isa: " . ref(connection) );
    Rex::Config->set_no_tty(1);
    my $cmd = qq{perl -e 'print STDERR "x" for (0..$count);'};
    Rex::Logger::info($cmd);
    my $out = run $cmd, sub {
        my ( $out, $err ) = @_;
        Rex::Logger::info( "stdout length: " . length($out) );
        Rex::Logger::info( "stderr length: " . length($err) );
    }, auto_die => 1;
};
1;
```

Call it like so:
```
$ rex -H localhost print_chars_to_stderr --count=64000
[2015-08-07 12:38:54] INFO - Running task print_chars_to_stderr on localhost
[2015-08-07 12:38:54] INFO - Connecting to localhost:22 (mitch)
[2015-08-07 12:38:54] INFO - Connected and authenticated to localhost.
[2015-08-07 12:38:54] INFO - Successfully authenticated on localhost.
[2015-08-07 12:38:54] INFO - connection isa: Rex::Interface::Connection::OpenSSH
[2015-08-07 12:38:54] INFO - perl -e 'print STDERR "x" for (0..64000);'
```

If it doesn't hang for you, increase the --count parameter and eventually it does. Interestingly, this doesn't happen when printing to STDOUT.

This problem presents itself only when set_no_tty=1. If set_no_tty=0, then it works fine

Can anyone else confirm this issue?

**Comment by ferki** [comment]
Confirmed. It would be nice if we could crosscheck somehow whether it's a bug in the Rex code or in Net::OpenSSH (or both).

**Comment by kablamo** [comment]
Confirmed also.  Works fine if I use ssh on the command line.  I'm leaning towards it being a Net::OpenSSH bug.  If I add debug statements around here:
https://github.com/RexOps/Rex/blob/master/lib/Rex/Interface/Exec/OpenSSH.pm#L38
I can see it hangs while trying to read from `$out_fh`.

I think the next step is to try to recreate this using just Net::OpenSSH and if thats possible, submitting it as a bug.  Shouldn't be super hard to do.   Just to be clear though, I'm not volunteering to do that.  :smile:  

https://metacpan.org/pod/Net::OpenSSH
https://rt.cpan.org/Public/Dist/Display.html?Name=Net-OpenSSH

**Comment by ferki** [comment]
I plan to take a deeper look at it together with the author of Net::OpenSSH at the upcoming Perl, SSH & Rex hackathon this Saturday at YAPC::EU ^^

**Comment by ferki** [comment]
Looks like this happens because the pipes attached to STDOUT and STDIN are being drained separetely instead of concurrently. So when the transfer window reaches its limit (64k on my machine) it stops the process to wait for the buffer to be drained before continuing.

**Comment by kablamo** [comment]
Ah, that makes sense.  But its not clear to me how to drain them concurrently.  Probably requires with lots of `select()` monkey business and other filehandle magic.  

I'm new to Net::OpenSSH, but my guess is we should outsource the filehandle magic to Net::OpenSSH by using `$ssh->pipe_in()` when people don't want a tty and `$ssh->open2pty()` when people do want a pty.  I don't see why anyone would want to do separate things with stdout vs stderr for an application like rex.  

That would not only solve this problem, but the much more important [bug](https://github.com/RexOps/Rex/pull/754) (imo) where rex's `run()` cmd can't stream output from commands ala tail -f.

FWIW I'm using `$ssh->open2pty()` for everything at work and thats been working well for me.

**Comment by ferki** [comment]
The [`_io3` implementation](https://github.com/salva/p5-Net-OpenSSH/blob/master/lib/Net/OpenSSH.pm#L1688-L1819) within `Net::OpenSSH` can be used as a reference. @salva (author of `Net::OpenSSH`) suggested copy-pasting it, and then reducing/simplifying it so it only does what Rex needs.

@kablamo: you're right, while dealing with `Net::OpenSSH` directly, it most probably would be possible to use another method to capture STDOUT/STDERR, instead of `open3` .

On the other hand, and thinking long-term, using `Net::SSH::Any` (tracked on #780) as a single interface towards any SSH implementation sounds like a good idea (so when it's stable, we can drop some ballast, and help upstream instead of replicating some stuff inside Rex).

**Comment by kablamo** [comment]
Oh right.  `Net::SSH::Any` doesn't support `open2pty()` or `pipe_in()`.  

I love the idea of `Net::SSH::Any` but I would argue the streaming output problem solved by using `open2pty()` and `pipe_in()` are much more important than the code simplification gain we get from using `Net::SSH::Any`.

**Comment by ferki** [comment]
@kablamo: of course, we're gonna fix this first! There's probably a long way till using `Net::SSH::Any` fully, but the idea is to add SSH-related functionality there (and/or in related modules), instead of in Rex. Sorry if it sounded confusing, I think we just found our first candidates :)

**Comment by ferki** [comment]
@mbroadhead, @kablamo, @krimdomu: it'd be nice if you could give a spin to my proposed fix for this on #820.

**Comment by mbroadhead** [comment]
@ferki: I just tested #820 on my Macbook and it works.

First, a run without your pull request. As you can see, it hung and I interrupted it:
```
Mitchells-MacBook-Pro:Rex mitch$ rex -f ~/tmp/Rexfile print_chars_to_stderr --count=128000
[2015-11-01 16:34:46] INFO - Running task print_chars_to_stderr on <local>
[2015-11-01 16:34:46] INFO - connection isa: Rex::Interface::Connection::Local
[2015-11-01 16:34:46] INFO - perl -e 'print STDERR "x" for (0..128000);'
^C
```

With your pull request:
```
Mitchells-MacBook-Pro:Rex mitch$ perl -Ilib bin/rex -f ~/tmp/Rexfile print_chars_to_stderr --count=128000
[2015-11-01 16:35:12] INFO - Running task print_chars_to_stderr on <local>
[2015-11-01 16:35:12] INFO - connection isa: Rex::Interface::Connection::Local
[2015-11-01 16:35:12] INFO - perl -e 'print STDERR "x" for (0..128000);'
[2015-11-01 16:35:12] INFO - stdout length: 0
[2015-11-01 16:35:12] INFO - stderr length: 128001
[2015-11-01 16:35:13] INFO - SUMMARY
[2015-11-01 16:35:13] INFO - All tasks successful on all hosts
```

**Comment by ferki** [comment]
We believe this issue was partially fixed on #820. Which unfortunately introduced some regressions (#887), which were addressed on #888 and #908. All of this work is available in the `1.4rc` branch as of now, and will be included in the next release.

Based on the above, I'm going to close this issus, but feel free to reopen if needed.
